File: Daily Questions/2463-minimum-total-distance-traveled/2463-minimum-total-distance-traveled.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define min(a, b) (((a) < (b)) ? (a) : (b))

/**
 * @brief Compare function for sorting integers in ascending order.
 *
 * @param a (const void*): Pointer to the first element to compare.
 * @param b (const void*): Pointer to the second element to compare.
 *
 * @return (int): Returns the difference between the two elements, determining their order.
 */
int cmpfunc(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

/**
 * @brief Compare function for sorting factories by position in ascending order.
 *
 * @param a (const void*): Pointer to the first factory's position.
 * @param b (const void*): Pointer to the second factory's position.
 *
 * @return (int): Returns the difference in positions between the two factories.
 */
int cmpfunc2(const void *a, const void *b) {
    return ((*(int **)a)[0] - (*(int **)b)[0]);
}

/**
 * @brief Calculate the minimum total distance that robots must travel to reach repairable factories.
 *
 * Uses dynamic programming to find the optimal assignment of robots to factories under the repair limits
 * of each factory, minimizing the travel distance.
 *
 * @param robot (int*): Array of robot positions.
 * @param robotSize (int): Number of robots in the robot array.
 * @param factory (int**): 2D array representing factories, where each factory has [position, limit].
 * @param factorySize (int): Number of factories in the factory array.
 * @param factoryColSize (int*): Array specifying the number of columns in each row of the factory array.
 *
 * @return (long long): Minimum total distance for all robots to be repaired.
 */
long long minimumTotalDistance(int* robot, int robotSize, int** factory, int factorySize, int* factoryColSize) {
    // Sort robots and factories for efficient assignment
    qsort(robot, robotSize, sizeof(int), cmpfunc);
    qsort(factory, factorySize, sizeof(int *), cmpfunc2);

    // Initialize dp array with maximum possible value (INT_MAX)
    long long dp[robotSize + 1][factorySize + 1];
    for (int i = 0; i <= robotSize; ++i) {
        for (int j = 0; j <= factorySize; ++j) {
            dp[i][j] = LLONG_MAX;
        }
    }
    dp[0][0] = 0;  // No robots, no factories, cost is zero.

    // Fill dp array to calculate minimum distances
    for (int j = 1; j <= factorySize; ++j) {
        int pos = factory[j - 1][0];    // Factory position
        int limit = factory[j - 1][1];  // Maximum robots the factory can repair

        for (int i = 0; i <= robotSize; ++i) {
            // Case 1: Skip the current factory
            dp[i][j] = dp[i][j - 1];

            // Case 2: Assign up to `limit` robots to this factory
            long long total_distance = 0;
            for (int k = 1; k <= min(i, limit); ++k) {
                total_distance += abs(robot[i - k] - pos);
                if (dp[i - k][j - 1] != LLONG_MAX) {
                    dp[i][j] = min(dp[i][j], dp[i - k][j - 1] + total_distance);
                }
            }
        }
    }

    return dp[robotSize][factorySize];
}

File: Difficulty/C. Hard/2463-minimum-total-distance-traveled/2463-minimum-total-distance-traveled.py
from typing import List
from functools import lru_cache

class Solution:
    """
    Calculates the minimum total distance traveled by robots to reach factories.
    """
    def minimumTotalDistance(self, robot: List[int], factory: List[List[int]]) -> int:
        """
        Determines the minimum total distance all robots need to travel to be repaired,
        given each factory's repair limits and robot/factory positions.

        Parameters:
        robot (List[int]): List of robot positions along the X-axis.
        factory (List[List[int]]): List of factories, where each factory is represented
                                   by [position, limit].

        Returns:
        int: Minimum total distance all robots need to travel for repairs.
        """
        # Sort robots and factories for efficient assignment
        robot.sort()
        factory.sort()

        n, m = len(robot), len(factory)

        @lru_cache(None)
        def dp(i: int, j: int) -> int:
            """
            Recursively calculates the minimum distance to assign robots to factories,
            utilizing memoization for efficiency.

            Parameters:
            i (int): Current index of the robot being assigned.
            j (int): Current index of the factory being considered.

            Returns:
            int: Minimum total travel distance for robots to be repaired up to index `i`.
            """
            if i == n:  # All robots assigned
                return 0
            if j == m:  # Out of factories to assign to
                return float('inf')

            # Option 1: Skip this factory
            res = dp(i, j + 1)

            # Option 2: Assign robots to the current factory within its limit
            total_distance = 0
            for k in range(min(factory[j][1], n - i)):
                total_distance += abs(robot[i + k] - factory[j][0])
                res = min(res, total_distance + dp(i + k + 1, j + 1))

            return res

        return dp(0, 0)
